crates/rust-analyzer/src/cli/load_cargo.rs

```diff
@@ -13,17 +13,27 @@ use vfs::{loader::Handle, AbsPath, AbsPathBuf};
 
 use crate::reload::{ProjectFolders, SourceRootConfig};
 
+pub fn load_workspace(
+    root: &Path,
+    config: &CargoConfig,
+    progress: &dyn Fn(String),
+) -> Result<ProjectWorkspace> {
+    let root = AbsPathBuf::assert(std::env::current_dir()?.join(root));
+    let root = ProjectManifest::discover_single(&root)?;
+
+    ProjectWorkspace::load(root, config, progress)
+}
```

Comment:
It feels like this function doesn't pull its weight? Let's maybe remove it altogether and let the caller call the three methods, or introduce a convenince wrapper:

```rust
pub fn analyze_cargo_workspace_at(
  root: &Path,
  cargo_config: &CargoConfig,
  load_config: &LoadCargoConfig,
  progress: &dyn Fn(String),
) -> Result<(AnalysisHost, vfs::Vfs)> {

}

pub fn analyze_workspace(
 ws: ProjectWorkspace,
 config: &LoadCargoConfig,
 progress: &dyn Fn(String),
) -> Result<(AnalysisHost, vfs::Vfs)> {
```

Reply:
Hah, I too was hesitant about adding it. I was unsure how you would feel about introducing quite a bit of redundant workspace discovery logic to the code. I have no objections against just removing it though.

Feel free to pick whatever works best for you and I'll make the necessary changes. :)

a) remove it altogether and let the caller call the three methods
b) remove it and introduce a convenience wrapper instead

Reply:
I'd left the judgement of what's best to you, as the code is in your icache, but not mine ) Both a or b would be fine by me!

Reply:
Cool. Does [this](https://github.com/rust-analyzer/rust-analyzer/pull/7690/commits/1a441682606dc6c5240e2516ac65717cc7f25ecf) work for you then?

```rust
pub struct LoadCargoConfig {
    pub load_out_dirs_from_check: bool,
    pub with_proc_macro: bool,
}

pub fn load_workspace_at(
    root: &Path,
    cargo_config: &CargoConfig,
    load_config: &LoadCargoConfig,
    progress: &dyn Fn(String),
) -> Result<(AnalysisHost, vfs::Vfs)> {
    // ...
}

pub fn load_workspace(
    ws: ProjectWorkspace,
    config: &LoadCargoConfig,
    progress: &dyn Fn(String),
) -> Result<(AnalysisHost, vfs::Vfs)> {
    // ...
}

fn load_crate_graph(
    crate_graph: CrateGraph,
    source_root_config: SourceRootConfig,
    vfs: &mut vfs::Vfs,
    receiver: &Receiver<vfs::loader::Message>,
) -> AnalysisHost {
    // ...
}
```